### PR TITLE
#1896 create new work package button

### PR DESCRIPTION
**Pull request by Aaryan1203**
## Changes

Created new work package button logic

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No linting errors / newline at end of file warnings
- [ ] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Screenshots of UI changes (see Screenshots section)
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] No `yarn.lock` changes (unless dependencies have changed)
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes # (issue #)

### CODE REVIEW

File: src/frontend/src/pages/WorkPackageForm/CreateWorkPackageCRForm.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  const currentUser = useCurrentUser();
+  const onSubmit = async (payload: CreateWorkPackageApiInputs) => {
+    const crPayload = {
+      submitter: currentUser,
+      wbsNum: validateWBS(wbsNum),
+      type: ChangeRequestType.Issue,
+      what: payload.name,
+      why: [{ explain: 'explination', type: ChangeRequestReason.Competition }],
+      proposedSolutions: [],
+      duration: payload.duration,
```
Comment (RChandler234) [inline src/frontend/src/pages/WorkPackageForm/CreateWorkPackageCRForm.tsx:30]: these all need to go in a workPackageProposedChanges object that is of type WorkPackageProposedChangesCreateArgs

**Review by RChandler234 — CHANGES_REQUESTED** [review]
If you look in the console, this is failing when you submit, I think you'll need to update the endpoint so that proposedSolutions can be an empty array if projectProposedChanges or workPackageProposedChanges are set

**Review by RChandler234 — CHANGES_REQUESTED** [review]

File: src/frontend/src/pages/WorkPackageForm/WorkPackageFormView.tsx [hunk 2]
```diff
@@ -16,38 +15,24 @@ import PageLayout from '../../components/PageLayout';
...
-    .required('CR ID is required')
-    .typeError('CR ID must be a number')
-    .integer('CR ID must be an integer')
-    .min(1, 'CR ID must be greater than or equal to 1')
-});
-
 interface WorkPackageFormViewProps {
   exitActiveMode: () => void;
-  mutateAsync: (data: WorkPackageApiInputs) => void;
+  mutateAsync: (data: CreateWorkPackageApiInputs) => void;
```
Comment (RChandler234) [inline src/frontend/src/pages/WorkPackageForm/WorkPackageFormView.tsx:28]: revert back to WorkPackageApiInputs

File: src/frontend/src/pages/WorkPackageForm/WorkPackageFormView.tsx [hunk 3]
```diff
@@ -132,15 +118,16 @@ const WorkPackageFormView: React.FC<WorkPackageFormViewProps> = ({
...
+        projectWbsNum: wbsElement.wbsNum,
         workPackageId: defaultValues?.workPackageId,
         userId,
         name,
         crId: parseInt(crId),
         startDate: transformDate(startDate),
         duration,
         blockedBy: blockedByWbsNums,
-        expectedActivities: createForm ? expectedActivities.map((activity) => activity.detail) : expectedActivities,
-        deliverables: createForm ? deliverables.map((deliverable) => deliverable.detail) : deliverables,
```
Comment (RChandler234) [inline src/frontend/src/pages/WorkPackageForm/WorkPackageFormView.tsx:143]: revert this change actually, we need the ids for the edit page

File: src/shared/src/types/change-request-types.ts [hunk 4]
```diff
@@ -129,9 +129,9 @@ export interface ProjectProposedChangesCreateArgs extends WBSProposedChangesCrea
...
   duration: number;
-  startDate: string;
-  stage: WorkPackageStage | null;
-  blockedBy: WbsNumber[];
+  startDate: Date;
+  stage: string;
+  blockedBy: string[];
   expectedActivities: string[];
   deliverables: string[];
 }
```
Comment (RChandler234) [inline src/shared/src/types/change-request-types.ts:137]: this is the big thing, WorkPackageProposedChangesCreateArgs shouldn't extend the WBS one. Delete the WBS one and make these create args this:
```
  name: string;
  duration: number;
  startDate: string;
  stage?: WorkPackageStage;
  blockedBy: WbsNumber[];
  expectedActivities: string[];
  deliverables: string[];
  projectLeadId?: number;
  projectManagerId?: number;
```

Comment (RChandler234) [reply src/shared/src/types/change-request-types.ts:137]: this is going to require you to update a couple places in the backend/ frontend to get the types (particularly undefined stuff) to play nice

Comment (RChandler234) [reply src/shared/src/types/change-request-types.ts:137]: you should also update the workPackageProposedChanges validation stuff to reflect this

**Review by RChandler234 — CHANGES_REQUESTED** [review]
This is going to be a lot of specific changes, for now don't try to understand everything all at once, just try to make these changes to get the wireup working, and ask me any questions afterwards

File: src/frontend/src/apis/work-packages.api.ts [hunk 5]
```diff
@@ -4,16 +4,16 @@
...
 
 export interface WorkPackageApiInputs {
   name: string;
-  startDate: String;
+  startDate: string;
   duration: number;
-  crId: number;
+  crId: number | undefined;
   stage: WorkPackageStage | null;
   blockedBy: WbsNumber[];
```
[the remark this answers is not in the document]
Comment (RChandler234) [reply src/frontend/src/apis/work-packages.api.ts:18]: @Aaryan1203

**Comment by Aaryan1203** [comment]
The create CR thing works fine now. When I try to edit it, this is the error I get. I think it has to do with the type of WorkPackageApiInputs as it no longer is expecting the id and details. 
Screenshot 2024-04-22 at 10 48 45 PM

File: src/backend/src/services/change-requests.services.ts [hunk 6]
```diff
@@ -636,7 +637,7 @@ export default class ChangeRequestsService {
         otherConstraints
       } = projectProposedChanges;
 
-      await validateProposedChangesFields(projectLeadId, projectManagerId, links);
+      await validateProposedChangesFields(projectLeadId, projectManagerId);
```
Comment (RChandler234) [inline src/backend/src/services/change-requests.services.ts:635]: we should still be validating this for projects, just pass in an empty array of links for wpProposedChanges and make projectLeadId/projectManagerId optional

File: src/shared/src/types/change-request-types.ts [hunk 7]
```diff
@@ -112,8 +112,8 @@ export interface ProposedSolutionCreateArgs {
 export interface WBSProposedChangesCreateArgs {
```
Comment (RChandler234) [inline src/shared/src/types/change-request-types.ts:112]: go ahead and delete this/ incorporate the necessary fields into ProjectProposedChangesCreateArgs

File: src/frontend/src/pages/WorkPackageForm/CreateWorkPackageCRForm.tsx [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+const CreateWorkPackageCRForm: React.FC = () => {
+  const query = useQuery();
+  const wbsNum = query.get('wbs');
+  const history = useHistory();
+
+  if (!wbsNum) throw new Error('WBS number not included in request.');
+
+  const { mutateAsync } = useCreateStandardChangeRequest();
+
+  const currentUser = useCurrentUser();
```
Comment (RChandler234) [inline src/frontend/src/pages/WorkPackageForm/CreateWorkPackageCRForm.tsx:21]: bitpick: move this up with the other hooks

File: src/frontend/src/layouts/PageTitle/PageTitle.tsx [hunk 9]
```diff
@@ -7,7 +7,7 @@ import { Typography, Box, Grid, useTheme } from '@mui/material';
 import { ReactElement, ReactNode } from 'react';
 
 interface PageTitleProps {
-  title: string;
+  title: ReactNode;
```
Comment (RChandler234) [inline src/frontend/src/layouts/PageTitle/PageTitle.tsx:10]: I'd like to avoid this if possible, it'd be cool if you can repurpose the chips prop to be more general, so it could accept smthn like an info bubble instead

File: src/frontend/src/pages/WorkPackageForm/WorkPackageFormView.tsx [hunk 10]
```diff
@@ -47,7 +33,8 @@ interface WorkPackageFormViewProps {
   leadOrManagerOptions: User[];
   blockedByOptions: { id: string; label: string }[];
   crId?: string;
-  createForm?: boolean;
+  formType: WPFormType;
+  schema: any;
```
Comment (RChandler234) [inline src/frontend/src/pages/WorkPackageForm/WorkPackageFormView.tsx:37]: let's constrain this type as much as we can, should at least be able to make it this: `ObjectSchema<any>`

**Review by RChandler234 — CHANGES_REQUESTED** [review]
There's a problem where adding a blocked by on edit is erroring in prisma, looking into rn

**Comment by RChandler234** [comment]
happens on the create cr for new work package page rn: 
image

**Review by RChandler234 — APPROVED** [review]
Amazing
